File: backend/routes/topup.py
from decimal import Decimal, InvalidOperation
from typing import Any, Dict

from flask import jsonify, request

from database import db
from models import Account, Transaction, User
from . import api_bp


def _parse_amount(value: Any) -> Decimal:
	try:
		return Decimal(str(value)).quantize(Decimal("0.01"))
	except (InvalidOperation, TypeError, ValueError):
		raise ValueError("Invalid amount")


def _get_or_create_account(user_id: int) -> Account:
	user = User.query.get(user_id)
	if not user:
		raise ValueError("User not found")
	if user.account:
+		return user.account
	account = Account(user_id=user.id)
	db.session.add(account)
	db.session.flush()
	return account


def _mock_validate_card(payload: Dict[str, Any]) -> None:
	card_number = str(payload.get("cardNumber", "")).strip()
	expiry_month = str(payload.get("expiryMonth", "")).strip()
	expiry_year = str(payload.get("expiryYear", "")).strip()
	cvv = str(payload.get("cvv", "")).strip()
	if not (card_number and expiry_month and expiry_year and cvv):
		raise ValueError("Missing card details")
	if len(card_number) < 12 or len(card_number) > 19:
		raise ValueError("Invalid card number")
	if not (cvv.isdigit() and 3 <= len(cvv) <= 4):
		raise ValueError("Invalid CVV")


def _mock_validate_bank(payload: Dict[str, Any]) -> None:
	bank_name = str(payload.get("bankName", "")).strip()
	account_number = str(payload.get("accountNumber", "")).strip()
	routing_number = str(payload.get("routingNumber", "")).strip()
	if not (bank_name and account_number and routing_number):
		raise ValueError("Missing bank transfer details")
	if not (len(routing_number) in (9,)):
		raise ValueError("Invalid routing number")


@api_bp.post("/topup/card")
def topup_card():
	try:
		data = request.get_json(silent=True) or {}
		user_id = int(data.get("userId"))
		amount = _parse_amount(data.get("amount"))
		if amount <= 0:
			raise ValueError("Amount must be positive")
		_mock_validate_card(data)

		account = _get_or_create_account(user_id)
		account.balance = (account.balance or Decimal("0.00")) + amount
		txn = Transaction(account_id=account.id, type="CARD_TOPUP", amount=amount, reference="CARD")
		db.session.add(txn)
		db.session.commit()
		return jsonify({"success": True, "balance": str(account.balance), "transactionId": txn.id}), 201
	except Exception as exc:
		db.session.rollback()
		return jsonify({"success": False, "error": str(exc)}), 400


@api_bp.post("/topup/bank")
def topup_bank():
	try:
		data = request.get_json(silent=True) or {}
		user_id = int(data.get("userId"))
		amount = _parse_amount(data.get("amount"))
		if amount <= 0:
			raise ValueError("Amount must be positive")
		_mock_validate_bank(data)

		account = _get_or_create_account(user_id)
		account.balance = (account.balance or Decimal("0.00")) + amount
		txn = Transaction(account_id=account.id, type="BANK_TOPUP", amount=amount, reference="BANK")
		db.session.add(txn)
		db.session.commit()
		return jsonify({"success": True, "balance": str(account.balance), "transactionId": txn.id}), 201
	except Exception as exc:
		db.session.rollback()
		return jsonify({"success": False, "error": str(exc)}), 400

File: server/src/routes/topup.js
import { Router } from 'express'
import { Account, Transaction, User } from '../models.js'
import { sequelize } from '../db.js'
import { generateAccountNumber, createNotification } from '../utils.js'

const router = Router()

function parseAmount(value) {
	const n = Number.parseFloat(String(value))
	if (!Number.isFinite(n) || n <= 0) throw new Error('Amount must be positive')
	return Number(n.toFixed(2))
}

async function getOrCreateAccount(userId, t) {
	const user = await User.findByPk(userId, { transaction: t })
	if (!user) throw new Error('User not found')
	let account = await Account.findOne({ where: { userId: user.id }, transaction: t, lock: t.LOCK.UPDATE })
	if (!account) {
		const accountNumber = await generateAccountNumber(t)
		account = await Account.create({ 
			userId: user.id,
			accountNumber,
			balance: '0.00',
			currency: 'BDT'
		}, { transaction: t })
	}
	return account
}

function validateCard(body) {
	const { cardNumber, expiryMonth, expiryYear, cvv } = body
	if (!cardNumber || !expiryMonth || !expiryYear || !cvv) throw new Error('Missing card details')
	if (String(cardNumber).length < 12 || String(cardNumber).length > 19) throw new Error('Invalid card number')
	if (!/^[0-9]{3,4}$/.test(String(cvv))) throw new Error('Invalid CVV')
}

function validateBank(body) {
	const { bankName, accountNumber, routingNumber } = body
	if (!bankName || !accountNumber || !routingNumber) throw new Error('Missing bank transfer details')
	if (!/^[0-9]{9}$/.test(String(routingNumber))) throw new Error('Invalid routing number')
}

function validatePrepaidCard(body) {
	const { cardNumber, pin } = body
	if (!cardNumber || !pin) throw new Error('Missing prepaid card details')
	if (String(cardNumber).length < 12 || String(cardNumber).length > 19) throw new Error('Invalid card number')
	if (String(pin).length < 4 || String(pin).length > 6) throw new Error('Invalid PIN')
}

router.post('/topup/card', async (req, res) => {
	try {
		const { userId, amount } = req.body || {}
		validateCard(req.body || {})
		const amt = parseAmount(amount)
		const result = await sequelize.transaction(async (t) => {
			const account = await getOrCreateAccount(Number(userId), t)
			const currentBalance = parseFloat(String(account.balance)) || 0
			const newBalance = currentBalance + amt
			account.balance = newBalance.toFixed(2)
			await account.save({ transaction: t })
			const txn = await Transaction.create({ accountId: account.id, type: 'CARD_TOPUP', amount: amt, reference: 'CARD' }, { transaction: t })
			
			// Create notification
			await createNotification(
				Number(userId),
				'Card Top-up Successful',
				`You have successfully added ৳${amt.toFixed(2)} to your account via card. Your new balance is ৳${newBalance.toFixed(2)}.`,
				'TOPUP',
				t
			)
			
			return { balance: newBalance.toFixed(2), transactionId: txn.id }
		})
		return res.status(201).json({ success: true, ...result })
	} catch (e) {
		return res.status(400).json({ success: false, error: e.message })
	}
})

router.post('/topup/bank', async (req, res) => {
	try {
		const { userId, amount } = req.body || {}
		validateBank(req.body || {})
		const amt = parseAmount(amount)
		const result = await sequelize.transaction(async (t) => {
			const account = await getOrCreateAccount(Number(userId), t)
			const currentBalance = parseFloat(String(account.balance)) || 0
			const newBalance = currentBalance + amt
			account.balance = newBalance.toFixed(2)
			await account.save({ transaction: t })
			const txn = await Transaction.create({ accountId: account.id, type: 'BANK_TOPUP', amount: amt, reference: 'BANK' }, { transaction: t })
			
			// Create notification
			await createNotification(
				Number(userId),
				'Bank Transfer Successful',
				`You have successfully added ৳${amt.toFixed(2)} to your account via bank transfer. Your new balance is ৳${newBalance.toFixed(2)}.`,
				'TOPUP',
				t
			)
			
			return { balance: newBalance.toFixed(2), transactionId: txn.id }
		})
		return res.status(201).json({ success: true, ...result })
	} catch (e) {
		return res.status(400).json({ success: false, error: e.message })
	}
})

router.post('/topup/prepaid', async (req, res) => {
	try {
		const { userId, amount, cardNumber, pin } = req.body || {}
		validatePrepaidCard(req.body || {})
		const amt = parseAmount(amount)
		const result = await sequelize.transaction(async (t) => {
			const account = await getOrCreateAccount(Number(userId), t)
			const currentBalance = parseFloat(String(account.balance)) || 0
			const newBalance = currentBalance + amt
			account.balance = newBalance.toFixed(2)
			await account.save({ transaction: t })
			const txn = await Transaction.create({ accountId: account.id, type: 'PREPAID_TOPUP', amount: amt, reference: 'PREPAID' }, { transaction: t })
			
			// Create notification
			await createNotification(
				Number(userId),
				'Prepaid Top-up Successful',
				`You have successfully added ৳${amt.toFixed(2)} to your account via prepaid card. Your new balance is ৳${newBalance.toFixed(2)}.`,
				'TOPUP',
				t
			)
			
			return { balance: newBalance.toFixed(2), transactionId: txn.id }
		})
		return res.status(201).json({ success: true, ...result })
	} catch (e) {
		return res.status(400).json({ success: false, error: e.message })
	}
})

export default router


